lib/active_reporting/function_adapters/mysql.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module ActiveReporting
+  module FunctionAdapters
+    module Mysql
+      # The list of available datetime values.
+      #
+      # @return [Array<Symbol>]
+      #
+      def self.datetime_precision_values
+        %i[microsecond millisecond second minute hour day week month quarter year decade
+           century millennium].freeze
+      end
+
+      # Whether the arg value is supported as a viable datetime
+      # value for performing datetime functions.
+      #
+      def self.valid_datetime_precision_value?(value)
+        datetime_precision_values.include?(value)
+      end
+
+      # Generate a date truncation statement in MySQL
+      # See https://www.postgresql.org/docs/10/static/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC
+      #
+      # @param [String, Symbol] datetime_precision_value
+      # @param [String] quoted_table_name
+      # @param [String] column_name
+      #
+      def self.date_truncate(datetime_precision_value, quoted_table_name, column_name)
+        create_date_trunc_function unless date_trunc_function_exists?
+        "ACTIVE_REPORTING_DATE_TRUNC('#{datetime_precision_value}', #{quoted_table_name}.#{column_name})"
+      end
+
+      def self.date_trunc_function_exists?
+        ActiveRecord::Base.connection.execute(
+          <<-SQL
+            SELECT ROUTINE_NAME
+            FROM INFORMATION_SCHEMA.ROUTINES
+            WHERE ROUTINE_TYPE="FUNCTION"
+            AND UCASE(ROUTINE_NAME) = 'ACTIVE_REPORTING_DATE_TRUNC';
+          SQL
+        ).first.present?
+      end
+
+      def self.create_date_trunc_function
+        ActiveRecord::Base.connection.execute(
+          <<-SQL
+            CREATE FUNCTION ACTIVE_REPORTING_DATE_TRUNC(field ENUM('microsecond', 'millisecond', 'second', 'minute', 'hour', 'day', 'week', 'month', 'quarter', 'year', 'decade', 'century', 'millennium'), source datetime(6))
+            RETURNS datetime(6)
+            DETERMINISTIC
+            BEGIN
+              -- Short-circut in the week, month, or year, since those computations are straightforward
+              IF field IN ('week') THEN RETURN STR_TO_DATE(CONCAT(YEARWEEK(source, 2), ' Sunday'), '%X%V %W'); END IF;
+              IF field IN ('month') THEN RETURN DATE_FORMAT(source, '%Y-%m-01'); END IF;
+              IF field IN ('year') THEN RETURN DATE_FORMAT(source, '%Y-01-01'); END IF;
```

Comment:
@girasquid I plugged in the code from the reference you provided. In theory (?) this would improve performance by jumping out of the MySQL function for these three values.